cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +286,194 @@ impl Validator {
+            ExprKind::Lit(_) => (
+                EntityDerefLevel { level: Some(0) }, //Literals can't be dereferenced
+                None,
+            ),
+            ExprKind::Var(_) => (max_allowed_level.clone(), None), //Roots start at `max_allowed_level`
+            ExprKind::Slot(_) => (EntityDerefLevel { level: Some(0) }, None), //Slot will be replaced by Entity literal so treat the same
+            ExprKind::Unknown(_) => (
+                EntityDerefLevel { level: Some(0) }, //Can't dereference an unknown
+                None,
+            ),
+            ExprKind::If {
+                test_expr,
+                then_expr,
+                else_expr,
+            } => {
+                let es = [test_expr, then_expr, else_expr];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::And { left, right } | ExprKind::Or { left, right } => {
+                let es = [left, right];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::UnaryApp { arg, .. } => {
+                self.check_entity_deref_level_helper(arg, max_allowed_level, policy_id)
+            }
+            // `In` operator decrements the LHS only
+            ExprKind::BinaryApp { op, arg1, arg2 } if op == &BinaryOp::In => {
+                let mut lhs =
+                    self.check_entity_deref_level_helper(arg1, max_allowed_level, policy_id);
+                let rhs = self.check_entity_deref_level_helper(arg2, max_allowed_level, policy_id);
+                lhs = (lhs.0.decrement(), lhs.1);
+                let new_level = Self::min(vec![lhs, rhs]).0;
+                if new_level.level < Some(0) {
+                    (
+                        new_level,
+                        Some(EntityDerefLevelViolation {
+                            source_loc: e.source_loc().cloned(),
+                            policy_id: policy_id.clone(),
+                            actual_level: new_level,
+                            allowed_level: max_allowed_level.clone(),
+                        }),
+                    )
+                } else {
+                    (new_level, None)
+                }
+            }
+            ExprKind::BinaryApp { arg1, arg2, .. } => {
+                let es = [arg1, arg2];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
```

Comment:
By returning the `min` here I think you get things like `principal == principal` has level `max_allowed_level` which feels wrong. Could that cause trouble anywhere? `(principal == principal).foo` won't ever be allowed due to typechecking, but maybe there's a another troublesome case.

Reply:
That feels right to me. Each node's `level` is the maximum number of times it could be dereferenced. I don't see how `principal == principal` could be dereferenced at all, but in a hypothetical world where it could be, I'd exect it to be dereferenceable `max_allowed_level` times.

Reply:
fair enough